README
This package contains a reimplementation of getopt(1).

PREFACE

Getopt(1) is a program to help shell scripts parse command-line parameters.
It is for example included in the util-linux distribution (upto version
2.7.1). But, there are some problems with that getopt(1) implementation, 
as listed in the 'BUGS' section of its man-page:

>BUGS
>     Whatever getopt(3) has.
>
>     Arguments containing white space or imbedded shell metacharacters gener-
>     ally will not survive intact;  this looks easy to fix but isn't.
>
>     The error message for an invalid option is identified as coming from
>     getopt rather than from the shell procedure containing the invocation of
>     getopt; this again is hard to fix.
>
>     The precise best way to use the set command to set the arguments without
>     disrupting the value(s) of shell options varies from one shell version to
>     another.

This implementation of getopt(1) is written to solve some of these problems,
while still staying (for all practical purposes) completely compatible with
other getopt(1) implementations.


INSTALLATION

Installation should be very easy. Just type 'make' to compile the sources.
It should compile cleanly, without any warnings, but even if it does not
you probably don't have to worry. You must use GNU Make and gcc, or you
will have to edit the Makefile. 

Type 'make install' to install the binary and the manual page. It installs 
by default in /usr/local/bin and /usr/local/man/man1, to install in /usr/bin 
and /usr/man/man1 try 'make install prefix=/usr'.

The example files can be installed in /usr/local/lib/getopt by calling
'make install_doc'.

If you do not trust the getopt(3) in your libc, or if you do not use a libc
with the GNU getopt(3) routines, you can use the gnu sources as provided 
in the gnu directory. Try 'make LIBCGETOPT=0'. Ignore any compile warnings.

You can check whether the new implementation of getopt is found first
in your path by calling 'bash test.bash'.


HIGHLIGHTS

It can do anything that the GNU getopt(3) routines can do.

It can cope with spaces and shell metacharacters within arguments.

It can parse long parameters.

It can shuffle parameters, so you can mix options and other parameters on
the command-line.

It can be easily identified as an enhanced getopt(1) from within shell scripts.

It can report parse errors as coming from the shell script.

It is fully compatible with other getopt(1) implementations.

COPYING

This program comes under the GNU general public licence version 2. See the 
file COPYING included in this package. Note that though you may freely
copy it, it is copyright (c) 1997-2014 by Frodo Looijaard 
<frodo@frodo.looijaard.name>.
Files in the gnu directory are from glibc-2.0.4: copyright (C) 1987, 88, 
89, 90, 91, 92, 93, 94, 95, 96, 97 Free Software Foundation, Inc.


DOWNLOADING

You can find the latest version of this program at:
  http://software.frodo.looijaard.name/getopt/


Changelog
20141124: Bumped up version number to 1.1.6
20141124: Properly quote backslashes for tcsh
20141124: Detect ambiguous long options properly
          Thanks to Klaus Wulff <dinw.klswlff@dfgh.net>
20141123: Create test harnass and tests
20141122: Update translations (sync with util-linux (git 20141120))
20141120: Update manual page (sync with util-linux (git 20141120))
20141120: Minor code changes (sync with util-linux (git 20141120))
20120812: Bumped up version number to 1.1.5
20120812: Update translations (sync with util-linux (2.21))
20120731: Fix build problems  (sync with util-linux (2.21))
20120719: Update manpage (sync with util-linux (2.21))
20120717: Remaining code restructuring (sync with util-linux (2.21))
20120717: Use xmalloc instead of our own (sync with util-linux (2.21))
20120716: Use static definitions (sync with util-linux (2.21))
20120715: Minor spacing issues resolved (sync with util-linux (2.21))
20120714: Minor code restructuring (sync with util-linux (2.21))
20051107: Bumped up version number to 1.1.4
20051107: Makefile: package target
20051107: Changed email and website to current ones
20051107: Fixed a few typos in the manpage (sync with util-linux (2.12r))
20030123: Bumped up version number to 1.1.3
20030123: Changed WITH_GETTEXT option to WITHOUT_GETTEXT
20030123: Renamed example programs to have the string getopt- prepended
20030123: Manpage: all minus signs are now escaped
20030123: Sync with util-linux (2.11y)
20010329: Bumped up version number to 1.1.2
20010329: Sync with util-linux
20010311: Bumped up version number to 1.1.1
20010311: Fixed stupid compilation error if NLS not selected
20000630: Bumped up version number to 1.1.0
20000630: Ported the Native Language Support patches from util-linux
          They are written by Arkadiusz Miśkiewicz <misiek@pld.org.pl>
20000627: Bumped up version number to 1.0.6
20000627: A few manpage tweaks
20000616: Introduced DESTDIR and directory creation in the Makefile.
19990716: Bumped up version number to 1.0.5
19990622: Made -u work
19990101: Bumped up version number to 1.0.4 (aka. 1.0.3a in util-linux)
19990101: Very small fix in main()
19990101: Updated util-linux Makefile
19980628: Bumped up version number to 1.0.3
19980628: Removed remaining incompatibility with tcsh-6.02 from parse.tcsh
19980626: Added separate Makefile for util-linux
19980625: Removed several bugs from parse.tcsh, partly thanks to Volker Kuhlmann
          (v.kuhlmann@elec.canterbury.ac.nz).
19980612: Removed reference to getopt in util-linux
19980611: Bumped up version number to 1.0.2
19980611: Fixed --version bug (was not available, though documented!)
19980611: Removed compiler warnings.
19980603: Bumped up version number to 1.0.1
19980603: Fixed sizeof() bug (should be strlen) in getopt.c, thanks to 
          Bob Proulx (rwp@hprwp.fc.hp.com).
19980505: Changed date field in LSM to proper syntax
19980505: Released version 1.0


nls.h
/*
    nls.h - NLS system interface
    Copyright (c) 2000-2014  Frodo Looijaard <frodo@frodo.looijaard.name>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/* This file is a small hack to keep both the stand-alone package and
 * the util-linux versions happy.
 * In the stand-alone version, we define NOT_UTIL_LINUX and use our own
 * nls.h definitions; in the util-linux version, we use the global
 * util-linux nls definitions
 */

#ifdef NOT_UTIL_LINUX

#ifndef GETOPT_NLS
#define GETOPT_NLS

#define PACKAGE "getopt"

#if WITHOUT_GETTEXT
#define _(Text) (Text)
#undef bindtextdomain
#define bindtextdomain(Domain,Directory) /* empty */
#undef textdomain
#define textdomain(Domain) /* empty */
#undef setlocale
#define setlocale(Category,Locale) /* empty */
#else /* not WITHOUT_GETTEXT */
#include <libintl.h>
#define _(Text) gettext (Text)
#endif /* WITHOUT_GETTEXT */

#endif /* def GETOPT_NLS */

#else /* not NOT_UTIL_LINUX */
#include "../lib/nls.h"
#endif /* NOT_UTIL_LINUX */


getopt.c
/*
 *  getopt.c - Enhanced implementation of BSD getopt(1)
 *  Copyright (c) 1997-2014 Frodo Looijaard <frodo@frodo.looijaard.name>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/*
 * Version 1.0-b4: Tue Sep 23 1997. First public release.
 * Version 1.0: Wed Nov 19 1997.
 *   Bumped up the version number to 1.0
 *   Fixed minor typo (CSH instead of TCSH)
 * Version 1.0.1: Tue Jun 3 1998
 *   Fixed sizeof instead of strlen bug
 *   Bumped up the version number to 1.0.1
 * Version 1.0.2: Thu Jun 11 1998 (not present)
 *   Fixed gcc-2.8.1 warnings
 *   Fixed --version/-V option (not present)
 * Version 1.0.5: Tue Jun 22 1999
 *   Make -u option work (not present)
 * Version 1.0.6: Tue Jun 27 2000
 *   No important changes
 * Version 1.1.0: Tue Jun 30 2000
 *   Added NLS support (partly written by Arkadiusz Miśkiewicz
 *     <misiek@pld.org.pl>)
 * Version 1.1.4: Mon Nov 7 2005
 *   Fixed a few type's in the manpage
 * Version 1.1.5: Sun Aug 12 2012
 *  Sync with util-linux-2.21, fixed build problems, many new translations
 * Version 1.1.6: Mon Nov 24 2014
 *  Sync with util-linux git 20141120, detect ambiguous long options, fix
 *  backslash problem in tcsh
 */

/* Exit codes:
 *   0) No errors, successful operation.
 *   1) getopt(3) returned an error.
 *   2) A problem with parameter parsing for getopt(1).
 *   3) Internal error, out of memory
 *   4) Returned for -T
 */
#define GETOPT_EXIT_CODE	1
#define PARAMETER_EXIT_CODE	2
#define XALLOC_EXIT_CODE	3
#define TEST_EXIT_CODE		4

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>

#if LIBCGETOPT
#include <getopt.h>
#else
#include "getopt.h"
#endif

#include "util-linux-compat.h"
#include "nls.h"
#include "xalloc.h"

/* NON_OPT is the code that is returned when a non-option is found in '+'
 * mode */
#define NON_OPT 1
/* LONG_OPT is the code that is returned when a long option is found. */
#define LONG_OPT 0

/* The shells recognized. */
typedef enum { BASH, TCSH } shell_t;


/* Some global variables that tells us how to parse. */
static shell_t shell = BASH;	/* The shell we generate output for. */
static int quiet_errors = 0;	/* 0 is not quiet. */
static int quiet_output = 0;	/* 0 is not quiet. */
static int quote = 1;		/* 1 is do quote. */

/* Allow changing which getopt is in use with function pointer */
int (*getopt_long_fp) (int argc, char *const *argv, const char *optstr,
		       const struct option * longopts, int *longindex);

/* Function prototypes */
static const char *normalize(const char *arg);
static int generate_output(char *argv[], int argc, const char *optstr,
			   const struct option *longopts);
static void parse_error(const char *message);
static void add_long_options(char *options);
static void add_longopt(const char *name, int has_arg);
static void print_help(void);
static void set_shell(const char *new_shell);

/*
 * This function 'normalizes' a single argument: it puts single quotes
 * around it and escapes other special characters. If quote is false, it
 * just returns its argument.
 *
 * Bash only needs special treatment for single quotes; tcsh also recognizes
 * exclamation marks within single quotes, and nukes whitespace. This
 * function returns a pointer to a buffer that is overwritten by each call.
 */
static const char *normalize(const char *arg)
{
	static char *BUFFER = NULL;
	const char *argptr = arg;
	char *bufptr;

	free(BUFFER);

	if (!quote) {
		/* Just copy arg */
		BUFFER = xmalloc(strlen(arg) + 1);
		strcpy(BUFFER, arg);
		return BUFFER;
	}

	/*
	 * Each character in arg may take up to four characters in the
	 * result: For a quote we need a closing quote, a backslash, a quote
	 * and an opening quote! We need also the global opening and closing
	 * quote, and one extra character for '\0'.
	 */
	BUFFER = xmalloc(strlen(arg) * 4 + 3);

	bufptr = BUFFER;
	*bufptr++ = '\'';

	while (*argptr) {
		if (*argptr == '\'') {
			/* Quote: replace it with: '\'' */
			*bufptr++ = '\'';
			*bufptr++ = '\\';
			*bufptr++ = '\'';
			*bufptr++ = '\'';
		} else if (shell == TCSH && *argptr == '\\') {
			/* Backslash: replace it with: '\\' */
			*bufptr++ = '\\';
			*bufptr++ = '\\';
		} else if (shell == TCSH && *argptr == '!') {
			/* Exclamation mark: replace it with: \! */
			*bufptr++ = '\'';
			*bufptr++ = '\\';
			*bufptr++ = '!';
			*bufptr++ = '\'';
		} else if (shell == TCSH && *argptr == '\n') {
			/* Newline: replace it with: \n */
			*bufptr++ = '\\';
			*bufptr++ = 'n';
		} else if (shell == TCSH && isspace(*argptr)) {
			/* Non-newline whitespace: replace it with \<ws> */
			*bufptr++ = '\'';
			*bufptr++ = '\\';
			*bufptr++ = *argptr;
			*bufptr++ = '\'';
		} else
			/* Just copy */
			*bufptr++ = *argptr;
		argptr++;
	}
	*bufptr++ = '\'';
	*bufptr++ = '\0';
	return BUFFER;
}

/*
 * Generate the output. argv[0] is the program name (used for reporting errors).
 * argv[1..] contains the options to be parsed. argc must be the number of
 * elements in argv (ie. 1 if there are no options, only the program name),
 * optstr must contain the short options, and longopts the long options.
 * Other settings are found in global variables.
 */
static int generate_output(char *argv[], int argc, const char *optstr,
			   const struct option *longopts)
{
	int exit_code = EXIT_SUCCESS;	/* Assume everything will be OK */
	int opt;
	int longindex;
	const char *charptr;

	if (quiet_errors)
		/* No error reporting from getopt(3) */
		opterr = 0;
	/* Reset getopt(3) */
	optind = 0;

	while ((opt =
		(getopt_long_fp(argc, argv, optstr, longopts, &longindex)))
	       != EOF)
		if (opt == '?' || opt == ':')
			exit_code = GETOPT_EXIT_CODE;
		else if (!quiet_output) {
			if (opt == LONG_OPT) {
				printf(" --%s", longopts[longindex].name);
				if (longopts[longindex].has_arg)
					printf(" %s", normalize(optarg ? optarg : ""));
			} else if (opt == NON_OPT)
				printf(" %s", normalize(optarg ? optarg : ""));
			else {
				printf(" -%c", opt);
				charptr = strchr(optstr, opt);
				if (charptr != NULL && *++charptr == ':')
					printf(" %s", normalize(optarg ? optarg : ""));
			}
		}

	if (!quiet_output) {
		printf(" --");
		while (optind < argc)
			printf(" %s", normalize(argv[optind++]));
		printf("\n");
	}
	return exit_code;
}

/*
 * Report an error when parsing getopt's own arguments. If message is NULL,
 * we already sent a message, we just exit with a helpful hint.
 */
static void __attribute__ ((__noreturn__)) parse_error(const char *message)
{
	if (message)
		warnx("%s", message);
	fprintf(stderr, _("Try `%s --help' for more information.\n"),
		program_invocation_short_name);
	exit(PARAMETER_EXIT_CODE);
}

static struct option *long_options = NULL;
static int long_options_length = 0;	/* Length of array */
static int long_options_nr = 0;		/* Nr of used elements in array */
#define LONG_OPTIONS_INCR 10
#define init_longopt() add_longopt(NULL,0)

/* Register a long option. The contents of name is copied. */
static void add_longopt(const char *name, int has_arg)
{
	char *tmp;
	static int flag;

	if (!name) {
		/* init */
		free(long_options);
		long_options = NULL;
		long_options_length = 0;
		long_options_nr = 0;
	}

	if (long_options_nr == long_options_length) {
		long_options_length += LONG_OPTIONS_INCR;
		long_options = xrealloc(long_options,
					sizeof(struct option) *
					long_options_length);
	}

	long_options[long_options_nr].name = NULL;
	long_options[long_options_nr].has_arg = 0;
	long_options[long_options_nr].flag = NULL;
	long_options[long_options_nr].val = 0;

	if (long_options_nr && name) {
		/* Not for init! */
		long_options[long_options_nr - 1].has_arg = has_arg;
		long_options[long_options_nr - 1].flag = &flag;
		long_options[long_options_nr - 1].val = long_options_nr;
		tmp = xmalloc(strlen(name) + 1);
		strcpy(tmp, name);
		long_options[long_options_nr - 1].name = tmp;
	}
	long_options_nr++;
}


/*
 * Register several long options. options is a string of long options,
 * separated by commas or whitespace. This nukes options!
 */
static void add_long_options(char *options)
{
	int arg_opt;
	char *tokptr = strtok(options, ", \t\n");
	while (tokptr) {
		arg_opt = no_argument;
		if (strlen(tokptr) > 0) {
			if (tokptr[strlen(tokptr) - 1] == ':') {
				if (tokptr[strlen(tokptr) - 2] == ':') {
					tokptr[strlen(tokptr) - 2] = '\0';
					arg_opt = optional_argument;
				} else {
					tokptr[strlen(tokptr) - 1] = '\0';
					arg_opt = required_argument;
				}
				if (strlen(tokptr) == 0)
					parse_error(_
						    ("empty long option after "
						     "-l or --long argument"));
			}
			add_longopt(tokptr, arg_opt);
		}
		tokptr = strtok(NULL, ", \t\n");
	}
}

static void set_shell(const char *new_shell)
{
	if (!strcmp(new_shell, "bash"))
		shell = BASH;
	else if (!strcmp(new_shell, "tcsh"))
		shell = TCSH;
	else if (!strcmp(new_shell, "sh"))
		shell = BASH;
	else if (!strcmp(new_shell, "csh"))
		shell = TCSH;
	else
		parse_error(_
			    ("unknown shell after -s or --shell argument"));
}

static void __attribute__ ((__noreturn__)) print_help(void)
{
	fputs(USAGE_HEADER, stderr);
	fprintf(stderr, _(
		" %1$s optstring parameters\n"
		" %1$s [options] [--] optstring parameters\n"
		" %1$s [options] -o|--options optstring [options] [--] parameters\n"),
		program_invocation_short_name);

	fputs(USAGE_OPTIONS, stderr);
	fputs(_(" -a, --alternative            Allow long options starting with single -\n"), stderr);
	fputs(_(" -l, --longoptions <longopts> Long options to be recognized\n"), stderr);
	fputs(_(" -n, --name <progname>        The name under which errors are reported\n"), stderr);
	fputs(_(" -o, --options <optstring>    Short options to be recognized\n"), stderr);
	fputs(_(" -q, --quiet                  Disable error reporting by getopt(3)\n"), stderr);
	fputs(_(" -Q, --quiet-output           No normal output\n"), stderr);
	fputs(_(" -s, --shell <shell>          Set shell quoting conventions\n"), stderr);
	fputs(_(" -T, --test                   Test for getopt(1) version\n"), stderr);
	fputs(_(" -u, --unquoted               Do not quote the output\n"), stderr);
	fputs(USAGE_SEPARATOR, stderr);
	fputs(USAGE_HELP, stderr);
	fputs(USAGE_VERSION, stderr);
	fprintf(stderr, USAGE_MAN_TAIL("getopt(1)"));
	exit(PARAMETER_EXIT_CODE);
}

int main(int argc, char *argv[])
{
	char *optstr = NULL;
	char *name = NULL;
	int opt;
	int compatible = 0;

	/* Stop scanning as soon as a non-option argument is found! */
	static const char *shortopts = "+ao:l:n:qQs:TuhV";
	static const struct option longopts[] = {
		{"options", required_argument, NULL, 'o'},
		{"longoptions", required_argument, NULL, 'l'},
		{"quiet", no_argument, NULL, 'q'},
		{"quiet-output", no_argument, NULL, 'Q'},
		{"shell", required_argument, NULL, 's'},
		{"test", no_argument, NULL, 'T'},
		{"unquoted", no_argument, NULL, 'u'},
		{"help", no_argument, NULL, 'h'},
		{"alternative", no_argument, NULL, 'a'},
		{"name", required_argument, NULL, 'n'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	init_longopt();
	getopt_long_fp = getopt_long;

	if (getenv("GETOPT_COMPATIBLE"))
		compatible = 1;

	if (argc == 1) {
		if (compatible) {
			/*
			 * For some reason, the original getopt gave no
			 * error when there were no arguments.
			 */
			printf(" --\n");
			return EXIT_SUCCESS;
		} else
			parse_error(_("missing optstring argument"));
	}

	if (argv[1][0] != '-' || compatible) {
		quote = 0;
		optstr = xmalloc(strlen(argv[1]) + 1);
		strcpy(optstr, argv[1] + strspn(argv[1], "-+"));
		argv[1] = argv[0];
		return generate_output(argv + 1, argc - 1, optstr,
				       long_options);
	}

	while ((opt =
		getopt_long(argc, argv, shortopts, longopts, NULL)) != EOF)
		switch (opt) {
		case 'a':
			getopt_long_fp = getopt_long_only;
			break;
		case 'h':
			print_help();
		case 'o':
			free(optstr);
			optstr = xmalloc(strlen(optarg) + 1);
			strcpy(optstr, optarg);
			break;
		case 'l':
			add_long_options(optarg);
			break;
		case 'n':
			free(name);
			name = xmalloc(strlen(optarg) + 1);
			strcpy(name, optarg);
			break;
		case 'q':
			quiet_errors = 1;
			break;
		case 'Q':
			quiet_output = 1;
			break;
		case 's':
			set_shell(optarg);
			break;
		case 'T':
			return TEST_EXIT_CODE;
		case 'u':
			quote = 0;
			break;
		case 'V':
			printf(UTIL_LINUX_VERSION);
			return EXIT_SUCCESS;
		case '?':
		case ':':
			parse_error(NULL);
		default:
			parse_error(_("internal error, contact the author."));
		}

	if (!optstr) {
		if (optind >= argc)
			parse_error(_("missing optstring argument"));
		else {
			optstr = xmalloc(strlen(argv[optind]) + 1);
			strcpy(optstr, argv[optind]);
			optind++;
		}
	}
	if (name)
		argv[optind - 1] = name;
	else
		argv[optind - 1] = argv[0];

	return generate_output(argv + optind - 1, argc-optind + 1,
			       optstr, long_options);
}
